### PR TITLE
Plotting examples

### PR DESCRIPTION
**Pull request by ivirshup**
Re #1604, @fidelram, I've got the sphinx extension working for examples (at least locally) here. You can check out a simple example for `sc.pl.embedding`.

Does this solve the problem you were running into?

<details>
<summary> Functions which need examples </summary>

Functions in `sc.pl`

- [ ] Embedding plots
    - [ ] `sc.pl.embedding`
    - [ ] `sc.pl.draw_graph`
    - [ ] `sc.pl.diffmap`
    - [ ] `sc.pl.pca`
    - [ ] `sc.pl.tsne`
    - [ ] `sc.pl.umap`
    - [ ] `sc.pl.spatial`
    - [x] `sc.pl.embedding_density`
- [ ] PCA specific
    - [ ] `sc.pl.pca_loadings`
    - [ ] `sc.pl.pca_overview`
    - [ ] `sc.pl.pca_scatter`
    - [ ] `sc.pl.pca_variance_ratio`
- [ ] PAGA
    - [ ] `sc.pl.paga`
    - [ ] `sc.pl.paga_adjacency`
    - [ ] `sc.pl.paga_compare`
    - [ ] `sc.pl.paga_path`
- [ ] DPT pseudotime
    - [ ] `sc.pl.dpt_groups_pseudotime`
    - [ ] `sc.pl.dpt_timeseries`
- [ ] Groupby
    - [x] `sc.pl.dotplot`
    - [ ] `sc.pl.matrixplot`
    - [ ] `sc.pl.clustermap`
    - [ ] `sc.pl.heatmap`
    - [ ] `sc.pl.dendrogram`
    - [ ] `sc.pl.stacked_violin`
    - [ ] `sc.pl.tracksplot`
    - [ ] `sc.pl.violin`
- [ ] Preprocessing
    - [ ] `sc.pl.filter_genes_dispersion`
    - [ ] `sc.pl.highest_expr_genes`
    - [ ] `sc.pl.highly_variable_genes`
- [ ] DE
    - [ ] `sc.pl.rank_genes_groups`
    - [ ] `sc.pl.rank_genes_groups_dotplot`
    - [ ] `sc.pl.rank_genes_groups_heatmap`
    - [ ] `sc.pl.rank_genes_groups_matrixplot`
    - [ ] `sc.pl.rank_genes_groups_stacked_violin`
    - [ ] `sc.pl.rank_genes_groups_tracksplot`
    - [ ] `sc.pl.rank_genes_groups_violin`
- [ ] Misc/ to be classified
    - [ ] `sc.pl.ranking`
    - [ ] `sc.pl.scatter`
    - [ ] `sc.pl.sim`
    - [ ] `sc.pl.correlation_matrix`
    - [ ] `sc.pl.matrix`
- [ ] Time series (???)
    - [ ] `sc.pl.timeseries`
    - [ ] `sc.pl.timeseries_as_heatmap`
    - [ ] `sc.pl.timeseries_subplot`

Other functions

- [x] `sc.pp.calculate_qc_metrics`
- [x] `sc.tl.embedding_density`
- [ ]  `sc.get.obs_df`
- [ ] `sc.get.rank_genes_groups_df`

</details>

### CODE REVIEW

**Comment by ivirshup** [comment]
Example of what the `dotplot` docs look like (locally) once this is working:

image

Does make that `bbox_inches="tight"` option seem important.

**Comment by fidelram** [comment]
@ivirshup  This looks great! Thanks.  The issue with the cropping of the labels is quite annoying and indeed `bbox_inches="tight"` should help. However, I don't think is nice to add that line for each example, but, on the other hand, if we add this to the scanpy code, many figures will be affected and would need to be updated.

The same issue also affects the test images which most are cropped.

**Comment by ivirshup** [comment]
Update: Nvm, I figured this out. Tt had to do with `qualname_overrides`, which I've updated.

<details>
<summary> Old problem </summary>

@flying-sheep, weird sphinx bug I'm running into:

* The readthedocs builds are failing after commit fc83ec3
* The error is:

```pytb
scanpy/scanpy/external/pp/_bbknn.py:docstring of scanpy.external.pp.bbknn:24: WARNING: py:class reference target not found: sklearn.neighbors._dist_metrics.DistanceMetric
```

* The error will still occur as long as this has been added:

```rst
    .. plot::
        :context: close-figs

        >>> import scanpy as sc
        >>> adata = sc.datasets.pbmc68k_reduced()
        >>> sc.tl.umap(adata)
```

* If I remove the `sc.tl.umap` line, the builds work fine, as `sklearn.neighbors.DistanceMetric` resolves and no warning is thrown.

For now, I'm going to remove the type annotation from `sc.external.pp.bbknn`, since it's causing the error. Any ideas why calling `sc.tl.umap` means `sklearn.neighbors._dist_metrics.DistanceMetric` can no longer resolve? I assume it has something to do with packages being imported in an unexpected order, but also this is real weird.

</details>

**Comment by ivirshup** [comment]
@fidelram 

I don't think making a broader change would be so bad right now, especially if it deals with this issue. However, if there was a way to just fix the issue here specifically that would also be good.

Separate issue these examples raise:

It looks like the colormap is changing at some point when it shouldn't:

image

Any idea what's up with this?

**Comment by ivirshup** [comment]
For this PR, I'm thinking I'm going to wait on a response from @flying-sheep to figure out what to do about the sklearn intersphinx problem, then clean it up for a merge. A new issue will be opened up with all the functions that need examples, and those can be added through separate PRs as the release progresses.

**Comment by fidelram** [comment]
@ivirshup the color change is caused because of the way the defaults are used. This is an issue that has to do with the visualisation of defaults in the documentation. 

When you add `.style(....)` as in the second case you report, the colormap changes to `winter` although it should stay as `Reds`. This happens because the method default value of `cmap` is `winter`.  Ideally, it should be `None` in which case it can be decided if the colormap was already set or if a default is needed.

However, setting the default of the colormap to `None` removes the default information from the sphinx documentation.

We had some discussions with @flying-sheep on how to address this that are still pending but ideas on how to address this are welcome.

**Comment by ivirshup** [comment]
Is there an open issue for this colormap problem? If so, could you link it? I'd like to track it for at least 1.8.

File: docs/dev/documentation.rst [hunk 1]
```diff
@@ -42,7 +42,21 @@ Some key points:
...
 * Check that the docs look like what you expect them too! It's easy to forget to add a reference to function, be sure it got added and looks right.
 
-Look at `sc.tl.louvain <https://github.com/theislab/scanpy/blob/a811fee0ef44fcaecbde0cad6336336bce649484/scanpy/tools/_louvain.py#L22-L90>`__ as an example for everything mentioned here:
+Look at `sc.tl.louvain <https://github.com/theislab/scanpy/blob/a811fee0ef44fcaecbde0cad6336336bce649484/scanpy/tools/_louvain.py#L22-L90>`__ as an example for everything mentioned here.
+
+Plots in docstrings
+~~~~~~~~~~~~~~~~~~~
+
+One of the most useful things you can include in a docstring is examples of how the function should be used.
+In the case of plotting functions, it's also possible to include show the resulting plots in the rendered docs.
```
Comment (fidelram) [inline docs/dev/documentation.rst]: Typo: 
".. also possible to include *and*? show the resulting plots in the rendered docs"

File: scanpy/plotting/_tools/scatterplots.py [hunk 2]
```diff
@@ -106,6 +106,17 @@ def embedding(
...
     -------
     If `show==False` a :class:`~matplotlib.axes.Axes` or a list of it.
+
+    Examples
+    --------
+
+    .. plot::
+        :context: close-figs
+
+        >>> import scanpy as sc
```
Comment (fidelram) [inline scanpy/plotting/_tools/scatterplots.py]: Do you want to also remove the ">>>" prefix here as in other examples?

**Comment by fidelram** [comment]
So far looking good. Maybe we can add to the TO-DO list, to fix issue with the 'cropped' labels.

File: scanpy/plotting/_tools/scatterplots.py (hunk 2 above)
Comment (ivirshup) [reply scanpy/plotting/_tools/scatterplots.py]: I was actually planning on removing this one, just because I think it's not a very useful example. I have some idea's about how we should handle example for the embedding plots (since they share so much, and are fairly complex) but I'll tag you in another issue about that.

File: docs/dev/documentation.rst (hunk 1 above)
Comment (ivirshup) [reply docs/dev/documentation.rst]: I think I meant 

> In the case of plotting functions, it's also possible to ~include~ show the resulting plots in the rendered docs.

But I've reworded the whole thing. It could probably use a rework, but the main point it to link to examples, and make notes about computation.

**Comment by ivirshup** [comment]
I think this is ready to merge. After that, I'll open the meta-issue about adding examples to all the listed functions.
